crypto/ecdsa/ecdsa.go

```diff
@@ -127,225 +123,146 @@ func (sig ThresholdSignature) RangeWhile(f func(hotstuff.ID) bool) {
+func (ec *ecdsaBase) Verify(signature consensus.QuorumSignature, message []byte) bool {
+	s, ok := signature.(MultiSignature)
+	if !ok {
+		ec.mods.Logger().Panicf("cannot verify signature of incompatible type %T (expected %T)", signature, s)
+	}
 
-		hash, ok := hashes[s.Signer()]
-		if !ok {
-			continue
-		}
+	l := signature.Participants().Len()
+	if l == 0 {
+		return false
+	}
 
-		sig, ok := s.(*Signature)
-		if !ok {
-			err = multierr.Append(err, fmt.Errorf("%w: %T", crypto.ErrWrongType, s))
-			continue
-		}
+	results := make(chan bool, l)
+	hash := sha256.Sum256(message)
 
-		// use the registered verifier instead of ourself to verify.
-		// this makes it possible for the signatureCache to work.
-		if ec.mods.Crypto().Verify(s, hash) {
-			thrSig[sig.signer] = sig
-		}
+	for _, sig := range s {
+		go func(sig *Signature, hash consensus.Hash) {
+			results <- ec.verifySingle(sig, hash)
+		}(sig, hash)
 	}
 
-	if len(thrSig) >= ec.mods.Configuration().QuorumSize() {
-		return thrSig, nil
+	valid := true
+	for range s {
+		if !<-results {
+			valid = false
+		}
 	}
 
-	return nil, multierr.Combine(crypto.ErrNotAQuorum, err)
+	return valid
 }
 
-// VerifyThresholdSignature verifies a threshold signature.
-func (ec *ecdsaCrypto) VerifyThresholdSignature(signature consensus.ThresholdSignature, hash consensus.Hash) bool {
-	sig, ok := signature.(ThresholdSignature)
+// BatchVerify verifies the given quorum signature against the batch of messages.
+func (ec *ecdsaBase) BatchVerify(signature consensus.QuorumSignature, batch map[hotstuff.ID][]byte) bool {
+	s, ok := signature.(MultiSignature)
 	if !ok {
-		return false
-	}
-	if len(sig) < ec.mods.Configuration().QuorumSize() {
-		return false
-	}
-	results := make(chan bool)
-	for _, pSig := range sig {
-		go func(sig *Signature) {
-			results <- ec.mods.Crypto().Verify(sig, hash)
-		}(pSig)
-	}
-	numVerified := 0
-	for range sig {
-		if <-results {
-			numVerified++
-		}
+		ec.mods.Logger().Panicf("cannot verify signature of incompatible type %T (expected %T)", signature, s)
 	}
-	return numVerified >= ec.mods.Configuration().QuorumSize()
-}
 
-// VerifyThresholdSignatureForMessageSet verifies a threshold signature against a set of message hashes.
-func (ec *ecdsaCrypto) VerifyThresholdSignatureForMessageSet(signature consensus.ThresholdSignature, hashes map[hotstuff.ID]consensus.Hash) bool {
-	ec.mods.Logger().Debug(hashes)
-	sig, ok := signature.(ThresholdSignature)
-	if !ok {
+	l := signature.Participants().Len()
+	if l == 0 {
 		return false
 	}
-	hashSet := make(map[consensus.Hash]struct{})
-	results := make(chan bool)
-	for id, hash := range hashes {
-		if _, ok := hashSet[hash]; ok {
-			return false
-		}
-		hashSet[hash] = struct{}{}
-		s, ok := sig[id]
+
+	results := make(chan bool, l)
+
+	set := make(map[string]struct{})
+	for id, sig := range s {
+		message, ok := batch[id]
 		if !ok {
 			return false
 		}
+		set[string(message)] = struct{}{}
```

Comment:
This might be costly for long messages. Perhaps it would be better to stringify the `hash` and use that as the map key?? It would be interesting to benchmark this. @hanish520

Reply:
Actually, we can just use the hash itself. No need to stringify